### PR TITLE
Fail the source record if the write fails

### PR DESCRIPTION
**Pull request by srkukarni**
### Motivation
If we have trouble writing the output, depending on the processing semantics, we will need to fail the source record.

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java [hunk 1]
```diff
@@ -268,7 +270,7 @@ public void open(Map<String, Object> config, SinkContext sinkContext) throws Exc
     }
 
     @Override
-    public void write(Record<T> record) throws Exception {
```
Comment (jerrypeng) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java]: why are removing "throws Exception" here?

Comment (srkukarni) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java]: because nothing ever throws any exception in the code

**Comment by srkukarni** [comment]
run integration tests
run java8 tests

**Comment by srkukarni** [comment]
run java8 tests
run integration tests

**Comment by srkukarni** [comment]
run integration tests

**Comment by srkukarni** [comment]
run java8 tests

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java (hunk 1 above)
Comment (jerrypeng) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java]: https://github.com/apache/pulsar/blob/master/pulsar-io/core/src/main/java/org/apache/pulsar/io/core/Sink.java#L44

**Comment by srkukarni** [comment]
run java8 tests

**Comment by merlimat** [comment]
Merged in 2.3.1 at 
a7223fc

**Pull request merged**